### ISSUE TITLE
Implement a REST thread in piservice

### ISSUE THREAD

**Issue by ellerychan** [brata-hsdc/brata.masterserver#13]
Implement a simple REST transaction to exercise Django and the database, and to lay the groundwork for implementing other REST transactions, which should mostly be pretty repetitive once we get the basic structure in place.

Assume that the team registration process will be similar this year to what it was last year.  Handle the incoming **Register** BRATA message.  See the [Master Server Web Service Interface](https://github.com/brata-hsdc/brata.masterserver/blob/hsdc2016/transition/ms_interface.md) document for transaction details.

**Comment by ellerychan** [comment]
Creating branch `first_rest_thread_issue13`.

**Comment by ellerychan** [comment]
This is the format of the register request from last year:  **`http://ms/m/brata-v00/register`**

I would like to change it around a bit, to be:  **`http://ms/piservice/register/brata-v00`**

`ms` will be replaced by the hostname or IP address of the actual MS server.

`piservice` is the name of the app that is handling it, although it does not have to be.  It would be shorter to call it `pisvc` or `piserv`, but it would be more fun to call it `servepi` or `serve_pi`.

#### Questions:
- Is it going to take an act of god to change the URL format?
- What is the significance of the `brata-v00` element?  Is it supposed to be identifying a particular version of the BRATA Android app?
- Will it be `brata-v01` this year?
- If we get a wrong version, should we send back some kind of compatibilty warning?

**Comment by jawaad-ahmad** [comment]
1. No objection to changing it around. We haven't deployed anything with which we need to be backward compatible, so we should be free to do what we want at this point because we control both sides of the interface.
   - Keep in mind that the BRATA code base is written to use the current API. If we want to change it, we need to communicate with the BRATA devs. (This would have been a good time to use the `@mention` if the lists were set up already.)
2. The `v00` was for the version of the API. Since we control both sides of the interface, we should be able to live without it at this time. I believe the `brata` was to indicate it was communication between BRATA and the MS so we could keep those messages segregated from those between the stations and the MS.
3. If we do decide to continue versioning, we would bump up to `v01` only if the API changed. I don't think it's necessary, but it would be good for us to keep it in mind so we don't end up breaking communications at the last minute due to changes.
4. Again since we're free to update both sides, we can negotiate the change and not worry about this; if we choose to maintain it, then yes, some sort of warning might be useful if the remote end gets reduced functionality by using the older API.

**Comment by ellerychan** [comment]
Turning off Django's **Cross Site Request Forgery (CSRF) Protection**.  It relies on a cookie being sent by the server in response to a GET request (which in a browser-based transaction would return a form containing the cookie). The POST response back to the server must include the cookie value in order to be accepted.

If the BRATA clients will support that, it can be turned back on.  However, having the feature activated also makes it difficult to test the REST API using simple `curl` commands.

Even turned off by default, it can still be activated for some transactions on a case-by-case basis.

A request that is rejected by the server's CSRF protection returns `403 Forbidden`.

Ref: https://docs.djangoproject.com/en/1.8/ref/csrf/

**Comment by jawaad-ahmad** [comment]
Will turning off CSRF cause us any problems down the road? Would it be better to fight the battle and leave CSRF on now rather than discover we need it later and have difficulty during a crunch?

Are you doing anything cross-site? I thought all projects inside the Django app would be accessible from the same host/port.

Also I recommend looking into `httpie`. It seems significantly more user-friendly than `curl` for what we need it to do.

**Comment by ellerychan** [comment]
I'm not super-clear on exactly what CSRF protects against.  Sounds like it prevents a malicious site from responding back to the server pretending to be you (the client).  Not likely in our scenario.  But I am also not sure how non-browser clients handle cookies, if at all.

In any case, I thought it would require the conversation to start with the client sending a **GET** request to the server, in order to get the CSRF token.  But I think the ICD makes it look like the client should start by sending the [`Register`](https://github.com/brata-hsdc/brata.masterserver/blob/hsdc2016/transition/ms_interface.md#register) **POST** message.  That seems nice and clean, and I'd hate to make it all messy and complex if the security is not necessary.

**Comment by jawaad-ahmad** [comment]
If it's only one simple **GET** request to get the token, that doesn't sound bad; however, it also doesn't seem like something we need to worry about at this point.

How about opening an issue to investigate further so we don't forget about it and cross-referencing it here? We can put it on Hold till we need it or decide to work it.

**Comment by ellerychan** [comment]
Roger. Will create an issue for it. 

-ellery-
(from miPhone)

**Comment by ellerychan** [comment]
Check this out.  This REST thread works.  (It's just one request and response.  Diagram created with yEd.)  Now I am trying to clean it up and add the database update part to it.  Then we will have a basic framework to work off of.  It should be quick to implement the specific messages, and then attention can be turned to other parts of the MS functionality like scoring and managing the db and putting up informative displays.

![ms_brata_registration_transition_diagram](https://cloud.githubusercontent.com/assets/6425759/9623467/07337ca2-510f-11e5-90bf-0d4470bb833a.png)

**Comment by jawaad-ahmad** [comment]
Looks good. We need a `message_version` more than the `brata_version`. The `brata_version` is nice to have, but the `message_version` is the one that really matters, and it's allowed to vary per message.

We had a `message_timestamp` too, didn't we?

I would not recommend overloading the version field to indicate the OS type; as long the MS gets the message, it processes it and provides a response.

If we need to know the OS type for some reason, we can ask for it via a dedicated field. Same for BRATA capabilities.

I like the idea of supporting BRATA capabilities in the future if needed. Are you thinking if someone brings in a tablet without GPS or a phone with camera disabled?

**Comment by ellerychan** [comment]
I thought that `brata_version` tells the MS that we are going to be communicating with a specific version of the message protocol, so all messages are implicitly in that dialect.  Seems like it could get really confusing to have individual message versions if they differ enough to require different message sequences.

**Comment by ellerychan** [comment]
> We had a message_timestamp too, didn't we?

I was having the MS add a timestamp to the message when it was received.

I wonder if we want the MS to include a timestamp or sequence number in all responses.

**Comment by jawaad-ahmad** [comment]
> I was having the MS add a timestamp to the message when it was received.

That's good info too, but by then we lose the time at which the message was generated, which could be seconds earlier assuming all the stations are synchronized to the same time source, which I thought was true. It might not make any difference now, but we might realize we need an application for it later.

As stated last year, if we don't need it and don't use it, we can leave it as zero, but we should at least design it into the interface now so we have it when needed.

> I wonder if we want the MS to include a timestamp or sequence number in all responses.

Both together would be ideal. The sequence number will be necessary if we ever need to troubleshoot dropped/lost messages, but it can be useful for an audit trail as well. Given how easy these are to implement, I recommend adding in up-front and then later removing if they get in the way for some reason.

> I thought that brata_version tells the MS that we are going to be communicating with a specific version of the message protocol, so all messages are implicitly in that dialect.

That's good information too, but it doesn't help with backward compatibility if we need to take it that far. Message versions and file format versions should never be tied to the software release, and each message's version number _should_ be allowed to vary independently. If we go for the next five years without changing the `register` API, there is no reason to bump up its version number just because we're now on a different release, but agreed it is nice to know what release of the software you're talking to.

> Seems like it could get really confusing to have individual message versions if they differ enough to require different message sequences.

No, it isn't more confusing, but it can end up being more to track if we ever need to take it that far, which usually isn't the case.

I don't see a need for it, so this too we can always leave at 0 (or maybe 1 now since we're changing every message from last year), but we should still build it in up-front rather than be forced to add it in later if we realize we need it then.

**Comment by ellerychan** [comment]
> That's good information too, but it doesn't help with backward compatibility if we need to take it that far. Message versions and file format versions should never be tied to the software release, and each message's version number should be allowed to vary independently. If we go for the next five years without changing the register API, there is no reason to bump up its version number just because we're now on a different release, but agreed it is nice to know what release of the software you're talking to.

Still seems too fine-grained to me, especially at the start.  When we work with an API, we get a new version number for the entire the library.  Some interface entry points have changed and some have not, yet we do not put a version number on each API function call.  I say we don't put a version in every message, and if we decide to later, then we increment the overall version and change the messages to include the individual message version.

**Comment by ellerychan** [comment]
Regarding the timestamp, and our difference in philosophy.  I tend to design like you and try to anticipate future needs, but sometimes it leads to bloat.  I am trying to be better and keep it minimal.  I think we should let it evolve.  This is not going into a ROM in an embedded system.  I think adding them when we need them is the safer approach.  Otherwise we run the risk of having extra fields get used in strange ways and then having weird legacy issues... although we probably will not need to be providing a lot of backward compatibility either.

However, I still believe that less code for people to read will lead to better adoption.

**Comment by jawaad-ahmad** [comment]
Very well. Let's start off slow.

**Comment by ellerychan** [comment]
We can either register **Team**s in advance, or let them register themselves.  I think we are talking about two different uses of the word _register_, however.

Preamble:  I've refactored the database design a little
- The `School` table is now the `Organization` table.  It will hold schools and HSDC at least.
- The `Mentor` table is now the `MSUser` table.  It will hold individuals, their contact information, and their login information.
1.  We _register_ the different entities with our system by entering their particulars into the database.  We add `Organizations` and `MSUsers`.  We could also add `Team` entries at that point.  Alternatively, we could let the Teams visit a web page and enter their own information.  This could be enabled in advance, or on competition day.
2.  This is in contrast to the `Register` REST message that the BRATA sends to the MS.  That signifies that the BRATA is present and ready to start the competition.

My inclination is to give the Teams the ability to register themselves with the MS.

One aspect that needs to be thought through is that permissions are attached to `Users`.  (An `MSUser` is a `User`.)  When a `Team` is registered, we will create an `MSUser` for that team.  So I think it will all work out nicely.  I will move ahead with this design.

Feedback?

**Comment by jawaad-ahmad** [comment]
> We could also add Team entries at that point. Alternatively, we could let
> the Teams visit a web page and enter their own information. This could be
> enabled in advance, or on competition day.

I vote to add in as many particulars as necessary in advance that cannot be
added the day of. (Is there any datum that fits this?) Known team names can
also be entered at this time.

Self registration can be turned on in the morning to minimize any chance of
mischief sooner, and teams can finish by signing-in and supplying final
info. If team names have already been entered, then this would merely check
that the team already exists.

This also captures attendance and can provide an opportunity to view/
correct/notify about errors previously given. ("Our team name is
'Megabytes', not 'Megabites'!")

I vote we call this _signing-in_ to avoid ambiguity with _registration_ as
previously mentioned.

> My inclination is to give the Teams the ability to register themselves
> with the MS.

Agreed. It lets the app communicate with the MS so teams find out if there
are any problems during the intro.

It reduces the amount of work we need to do to set up.

> One aspect that needs to be thought through is that permissions are
> attached to Users. (An MSUser is a User.) When a Team is registered, we
> will create an MSUser for that team. So I think it will all work out nicely.

What exactly needs to be thought through here?

**Comment by ellerychan** [comment]
Sounds like we are pretty much in concurrence.  Moving on.

**Comment by ellerychan** [comment]
@jawaad-ahmad, @DragonmasterJ:  I've pretty much completed a couple of initial MS threads (use cases?), so we should probably meet and sync up sometime soon.  I want to make sure I'm not going off into the weeds.  I'm letting you know now, in advance of completing this and issuing the Pull Request, so you can find a time we can meet.  Or we could try to do it online with Skype and screen sharing if you want to try something like that.  I've never done it before.

If you think I'm on the right track, then I will do some cleaning up and refactoring to try to make it a little more _cookbook-ish_, so that it will be easier to extend with other messages and screens.

The thread that this issue pertains to is the REST thread for the BRATA `register` message handled by **piservice**.  I also added some other infrastructure to help populate the database.  That work should really be done under Issue #11, but I needed it here.  I added basic screens to add `Organizations`, `Users`, and `Teams`, and a _check-in_ screen for Teams on competition day.

All the screens are rudimentary at the moment, and can use quite a bit of enhancement.  I'm opening up Issue #22 for that.

**Comment by jawaad-ahmad** [comment]
I'm fine meeting to go over this. Would you like come to our campus this time? We have large monitors in the rooms of the new building, plus it's probably more likely we can grab @jaron42 while he's there rather than asking him to drive up to your place.

**Comment by ellerychan** [comment]
Sure.  We just have to work around the after-hours security issues.

**Comment by ellerychan** [comment]
@jawaad-ahmad, @DragonmasterJ:  I made a condensed messaging document ([ms_message_summary.md](https://github.com/brata-hsdc/brata.masterserver/blob/first_rest_thread_issue13/transition/ms_message_summary.md)) so I can see more easily what is going on -- sort of a cheat-sheet.  It does is not meant to replace the **Master Server Web Service Interface** document ([ms_interface.md](https://github.com/brata-hsdc/brata.masterserver/blob/hsdc2016/transition/ms_interface.md)).

I would really like to remove all the message version information from the URLs (and probably from the JSON, too).  I would prefer the MS and device negotiate up front what message versions they will use to communicate.  It would require that a piece of state (the message version) would need to be saved for the session on each end of the communication channel, if the device cared about supporting multiple versions of the protocol:  if a device only supports one version of the protocol, then it either goes ahead or just gives up.

Removing the version info would make very nice, very readable URLs.

**Comment by jawaad-ahmad** [comment]
>  I made a condensed messaging document...

I like the condensed document. Too bad we have to duplicate some info in both places. It would be nice to link the messages from the table rows to jump to the appropriate section of the ms_interface.md document, but that's probably asking too much. I doubt we can do it with the docs in the repo rather than the wiki, and I don't think GitHub Flavored Markdown supports it.

> I would really like to remove all the message version information from the URLs (and probably from the JSON, too).

There are no plans to inter-operate with multiple versions; everything will run the same version everywhere, so I'm fine with entirely stripping them out. I don't think there's any need to negotiate anything at this point, so no need to save any message version state anywhere at this time either.

The only time I can think of a possible problem is if someone has the BRATA zip file left over from last year for some reason and tries to run it against this year's Master Server, maybe we can keep some version check at the initial `register` call and leave it at that. On the other hand, if we're changing all the URLs around anyway, then last year's BRATA won't be able to reach the Master Server anyway, so I'm fine stripping out all version checks this year entirely.

If we keep the same URL format next year, then we'll probably want to start adding the check next year.

**Comment by ellerychan** [comment]
I like that answer.

Last year's BRATA and Stations won't be able to talk to this year's MS.  They will get 404s trying to access last year's URLs.  We could catch those URLs easily enough if we wanted to, but I can't think of a good reason at the moment.  If we want to later, we can add it.

**Comment by ellerychan** [comment]
Next question:  What should we use for a unique Station ID?  Last year we gave them hostnames, but I wonder if it would be better to identify them by their MAC address.  However, even that can change if their network connection changes.  That would not be a problem unless a Pi has two interfaces, like if it was hardwired, but also had a WiFi dongle both connected to the same network.  If we assume one network connection, then IP address would work also, and is probably a lot easier to dig out.

I envision the Join going like this:

```
RPi Station  ----- Join host=192.168.1.5, station_type=CTS, pi_type=BPlus -----> MS
RPi Station <--------------- station_instance=2, station_id=1234 --------------  MS
```

Then the station can use `station_id` to identify itself in future communication.  It probably doesn't
need to know that it is instance 2, but it could set its hostname to CTS02 maybe.

So I guess I am recommending just using the IP address.  A limitation of this is that a station could only perform one function.  If we extended it slightly, a station could pretend to be multiple stations.  Perhaps we should leave that for an enhancement.

**Comment by jawaad-ahmad** [comment]
Have you thought about using part or all of the serial number?

http://raspberrypi.stackexchange.com/questions/2086/how-do-i-get-the-serial-number

**Comment by ellerychan** [comment]
I'll check it out.  Thanks.

**Comment by ellerychan** [comment]
@jaron42, @jawaad-ahmad, @DragonmasterJ: I'm going to open a Pull Request for this issue because I have a bunch of code to deliver that goes well beyond the original scope of the issue.

After that, I'll look for another issue where it is appropriate to continue development.

I still would like to get together to give you a live demo of what I have put together so far.  I have a bunch of ideas in my head, but you probably have some different ones, and I would like to hear them.  We can do it at my place or your place.

If you plan on scrutinizing the Pull Request, then I recommend we have the live session first so you can get a handle on the code.  It will make a lot more sense.  However, if you want to just rubber-stamp it so I can push it to the master (hsdc2016) branch, that is safe also, since there are no other merges to it, and all this code will get heavy modifications going forward anyway.

**Comment by jawaad-ahmad** [comment]
Two times I've observed merging early prior to conducting reviews:
- Get something stood up at the beginning.
- Get last-minute fixes in just before delivery.

In both cases, the mentality I've often observed is to go ahead and do what needs to be done and we can catch up on the paperwork later.

Since we're still in the initial development of the MS, I propose going ahead with the pull request and then going forward merging to `hsdc2016` as soon as you need to do so.

We can leave the pull request open for several days or longer for any additional discussion and then incorporate that in afterward.

I'm fine with the live demo. If you still need questions answered from @jaron42 then I propose meeting here since it'll be easier to get a hold of him in person that way.

**Comment by ellerychan** [comment]
> I'm fine with the live demo. If you still need questions answered from @jaron42 then I propose meeting here since it'll be easier to get a hold of him in person that way.

Let's do it.  I'm available any evening but Tuesday next week.  Wednesday would be best.  Thursday not so great but doable if it's the only time.  Someone on your end will have to arrange it and let me know.  I think 1.5 to 2 hours would be sufficient to cover everything and allow for some brainstorming.  We can go longer if you want.

**Comment by ellerychan** [comment]
@DragonmasterJ:  Can you tell me the names of all the station types that will be in the competition?  I only know **Return To Earth (RTE)**.

**Comment by DragonmasterJ** [comment]
@ellerychan:
I know we've discussed this through email and in person, but I wanted to document it here in writing as well. I'll take a look at any documents we have mentioning the station and make sure the names are updated.

The stations are:
Return to Earth (RTE or Return)
Secure
Dock
Launch

**Comment by jawaad-ahmad** [comment]
I recall the Launch challenge having no station; has this changed?

Dock could have a dedicated station or could be software running on
anything with a monitor connected. Have we decided what we are doing for it?

**Comment by ellerychan** [comment]
@jaron42:  I am implementing the old BRATA message formats alongside the new ones.  I figured I wouldn't throw away the new stuff.  I don't see how some of the old messages functioned, from looking at the docs (`Submit`, `Start Challenge`).  I'm guessing the docs are wrong.  I suppose I could look at the old PHP code.

**Comment by jawaad-ahmad** [comment]
Anything left to do for this, or can we mark this as resolved?

**Comment by ellerychan** [comment]
This can be closed.  It was merged into hsdc2016 in Pull Request #27.
